# Dzień piąty DWthon 👋 - Hack Outside the Box 

### Celem jest poznanie, czym jest tak zwany "aha moment"!

Czy wiesz co to jest "aha-moment"? To jest taki moment, kiedy mówisz... "aha" 😁. Czyli coś Cię olśniło i rzeczywistość (a przynajmniej jej kawałek) stał się bardziej dla Ciebie zrozumiały.

O "aha-momencie" można długo mówić, lepiej podam Ci kilka przykładów.


### Facebook (7 przyjaciół w 10 dni)
Jeśli użytkownik dodaje co najmniej **10 znajomych** w ciągu **pierwszych 7 dni** od rejestracji, wtedy ten użytkownik będzie bardziej zaangażowany i na dłużej zostanie w Facebook.

### Dropbox (1 plik, 1 folder, 1 urządzenie)
Jeśli użytkownik doda co najmniej jeden plik, jeden folder i zsynchronizuje się z co najmniej jednym urządzeniem, to wtedy ten użytkownik zostanie na dłużej.

### LinkedIn - X połączeń w Y dni


To są proste reguły, które potrafią opisać najważniejsze, co ma się wydarzyć w firmie, aby firma rosła. Innymi słowy to jest stosowanie wprost reguły 80/20.


Trudność polega na tym, że aby znaleźć tę prostą regułę, to... no właśnie, jak można ją znaleźć? Przecież takich prostych reguł może być tysiące lub nawet miliony, zabraknie nam czasu, aby je wszystkie sprawdzić. Tutaj na pomoc nam może przyjść `Machine Learning`.

### Gdzie zadawać pytania ❓ 

Jeśli napotkasz trudności podczas wykonywania zadań z tego notebooka, to koniecznie napisz w kanale [dwthon_day5](https://bit.ly/3qEBEMv)

Pamiętaj, aby szczegółowo doprecyzować, z czym masz problem. Najlepiej wrzuć screen z kodem swoim lub błędem, który widzisz i napisz, którego zadania dotyczy :)

In [1]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/EeFJLSS3_x0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

### Krok po kroku 

Nagrałem dla Ciebie także materiał wideo "krok po kroku". W wideo poniżej znajduje się dokładnie to, co w tym notebooku tylko tłumaczę wszystko, aby ułatwić Ci pracę i zrozumienie zagadnień i zadań :) Obejrzyj, jeśli potrzebujesz takiego dodatkowego wsparcia. 

In [2]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/7UTzlVkK17k" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Biblioteki
Potrzebujemy `pandas` i jeszcze trochę więcej innych bibliotek. 
* `DecisionTreeClassifier` - do trenowania modelu
* `sklearn.model_selection` walidacja modelu
* `gc` do sprzątania w pamięci RAM

Dodatkowo użyjemy nowego bardziej zaawansowanego modelu `xgboost` oraz biblioteki `eli5` do podglądania ważności cech.
`from collections import Counter` przyda nam się do wygodnego zliczenia liczby wystąpień (np. ile razy klient kupił produkt x).

In [3]:
import pandas as pd
import numpy as np

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
import xgboost as xgb
import eli5 
from collections import Counter

import gc

## Wczytajmy dane

In [4]:
df = pd.read_hdf("../input/data.h5")

print(df.shape)
df.sample(5)

(820906, 9)


order_id  customer_id  product_id  quantity  price_unit  price_total  \
471949     30256          700         711        10         635         6350   
486421     31368         1717        1312         4         675         2700   
842           55           39         495         5         165          825   
613183     40243          211        1506        12         145         1740   
33724       2441         1082        2372        12         125         1500   

        country_id          order_date  is_canceled  
471949           0 2010-02-16 14:49:00        False  
486421           0 2010-03-03 16:49:00        False  
842              0 2010-12-01 12:23:00        False  
613183           0 2010-07-09 10:26:00        False  
33724            0 2011-01-11 15:13:00        False

## Przygotowujemy dane


To jest kod z poprzedniej lekcji. Jedynie poukładamy kod w funkcji, aby łatwiej było nam eksperymentować później.

In [5]:
df_customers = (
    df[ ["price_total", "customer_id"] ]
    .groupby("customer_id")
    .agg("sum")
    .reset_index()
    .sort_values(by="price_total", ascending=False)
    .rename(columns={"price_total": "customer_price_total"})
)


df_customers["cumsum"] = df_customers["customer_price_total"].cumsum()
value_80prc = int(df["price_total"].sum() * 0.8)
df_customers["most_revenue_customer"] = df_customers["cumsum"] < value_80prc


top_customers = set(df_customers[ df_customers["most_revenue_customer"] ]["customer_id"].unique())

del df_customers
gc.collect()

def feature_engineering(df):
    df_customers = (
        df
        .groupby("customer_id")
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_products=("product_id", lambda x: len(set(x))),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_customers


def get_feats(df_customers, black_list=["most_revenue_customer"]):
    feats = list(df_customers.select_dtypes([np.number, bool]).columns)
    return [x for x in feats if x not in black_list]

def get_X_y(df_customers, top_customers, feats):
    df_customers["most_revenue_customer"] = df_customers["customer_id"].map(lambda x: x in top_customers)
    
    X = df_customers[feats].values
    y = df_customers["most_revenue_customer"].values
    
    return X, y


def train_and_get_scores(model, X, y, scoring="accuracy", cv=5):

    scores = cross_val_score(model, X, y, scoring=scoring, cv=cv)
    return np.mean(scores), np.std(scores)

## Model 1

☝️ Dokładnie to, co zrobiliśmy ostatnio.

In [6]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers)
X, y = get_X_y(df_customers, top_customers, feats)
model = DecisionTreeClassifier(max_depth=5)

train_and_get_scores(model, X, y)

(0.9996598639455783, 0.0004165798882284457)

Mamy wynik 99.9%.

Można się cieszyć, ale... tak naprawdę opowiem Ci o jednym mechanizmie, który regularnie stosuję. Nazywam go wahadło.

Idea polega na tym, aby najpierw być dużym optymistą i bardzo upraszczać rzeczywistość na początku. Następnie, jak już mamy dobry wynik, to wchodzimy w "zupełnie inne buty". Wchodzimy w rolę aktywnego "krytyka", który próbuje zrobić wszystko, aby udowodnić, że to, co zrobiliśmy nie ma sensu. Powtarzamy tak w kółko. Robimy kilka iteracji :)  

Najważniejsze jest to, że dzięki przełączaniu się pomiędzy rolami, można bardzo szybko biec od przodu, ale z drugiej strony mieć poczucie, że to nadal ma sens!

## 🧠 Włączmy myślenie krytycznie

Na ten moment wykonaliśmy pierwszy krok, czyli optymistyczny i dostaliśmy wynik 99%. Bardzo dobrze, ale... Problem polega na tym, że `price_unit_total`jest zbyt mocną cechą. Czyli ta cecha wprost mówi, czy klient należy do segmentu `top_customers` czy nie. Dlatego teraz musimy co najmniej tę cechę wyłączyć i sprawdzić, jak zachowa się model.


Wystarczy dodać `sum_price_total` do listy, którą będziemy ignorować. 

Zobacz 👀.


## Model 2

In [7]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = DecisionTreeClassifier(max_depth=5)

train_and_get_scores(model, X, y)

(0.7722420031842525, 0.2652334198347648)

Jak widzisz, jakość modelu nam spadła do 77%. Z jednej strony smutna sprawa 😂, ale z drugiej strony używamy modelu ML, aby znajdował mniej oczywiste zależności. Bo te oczywiste człowiek sam łatwo znajdzie sam 🤦‍♂️.


### Myślenie optymistyczne
To teraz wróćmy znów do bycia optymistą. Pamiętasz na czym polega zasada wahadła? Wahamy się tam i z powrotem ;) 


Można zmienić model na bardziej zaawansowany i to powinno poprawić wynik. Użyjmy `xgboost`. Ta nazwa może Ci nic nie mówić i to jest ok na tym etapie 👌.

## Model 3

Zwróć uwagę, że zmieniła się ta linijka: `model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)`.

In [8]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)

train_and_get_scores(model, X, y)

(0.8106769431176726, 0.08696513375768444)

## Mamy wynik 81%.

Spróbujmy zrobić pierwszą interpretację modelu i zobaczyć, co wg niego jest ważne.

In [9]:
model.fit(X, y)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=50, n_jobs=0, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [10]:
eli5.show_weights(model, feature_names=feats)

### Wstępnie wnioski:
#### TOP3 
- `sum_quantity` brzmi jako najważniejsza cecha*
- `count_orders`
- `sum_price_unit`

Wyglada na to, że cecha `count_unq_countries` jest mało przydatna.


☝️ Trzeba uważać na to, co jest "ważne", bo doświadczenie pokazuje, że to może być dość zmienne i w tej zmienności trzeba nauczyć się poruszać.


## Następny krok

Możemy teraz trochę pofilozofować nad tymi kilkoma cechami, ale to jest mało istotne. Te cechy same w sobie są dość oczywiste. Nie ma poczucia "dużej" wartości dodanej. 

Ciekawostką okazało się, że `sum_quantity` jest istotna. Tylko tutaj też nie ma dużego zaskoczenia.


Wygenerujmy zatem więcej cech, niech model znajdzie na co warto zwrócić naszą uwagę :) 


## Produkty


Możemy dodać dla każdego produktu osobną kolumnę i jako wartość w tej kolumnie dodać, ile produktów już było kupionych (czyli suma). W ten sposób chcemy znaleźć, które produkty są "najciekawsze", warte tego, aby zwrócić na nie uwagę.

☝️ Oczywiście możemy sprawdzić każdy produkt manualnie, ale mamy ich ponad 3.8k, może to nam trochę zająć czasu. Użyjmy lepiej do tego ML, będzie szybciej, sprawniej i efektywniej.

Tylko najpierw trzeba przygotować dane. Zmieńmy naszą funkcję `feature_engineering`. 

In [11]:
def feature_engineering(df):
    
    def counter(vals):
        cntr = Counter()
        cntr.update(vals)
        return cntr
    
    df_customers = (
        df
        .groupby("customer_id")
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_products=("product_id", lambda x: len(set(x))),
            count_by_products=("product_id", lambda x:  counter(x) ),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_customers


df_customers  = feature_engineering(df)
df_customers.head()

customer_id  count_orders  count_unq_products  \
0            0           159                  93   
1            1            40                 184   
2            2            30                 139   
3            3             9                  40   
4            4            14                  11   

                                   count_by_products  sum_quantity  \
0  {0: 89, 1: 73, 2: 69, 3: 68, 4: 32, 5: 36, 6: ...         22976   
1  {8: 8, 9: 4, 10: 4, 11: 4, 12: 3, 13: 15, 14: ...          3514   
2  {25: 9, 26: 7, 27: 7, 28: 3, 29: 4, 30: 10, 31...          8971   
3  {44: 7, 1781: 1, 1179: 1, 112: 1, 508: 3, 581:...          1108   
4  {54: 14, 1496: 2, 2288: 1, 2315: 1, 3432: 1, 4...           390   

   sum_price_unit  sum_price_total  count_unq_countries  prob_canceled  
0          895827          5746671                    1       0.010966  
1          205962           910113                    1       0.097727  
2           95500          1337464                    1       0.009050  
3           18626           226381                    1       0.000000  
4           20028           346978                    1       0.307692

Jak widzisz pojawiła się nowa kolumna `count_by_products`. W środku jest słownik. Zobaczmy dokładniej 👀.

In [12]:
df_customers["count_by_products"]

0       {0: 89, 1: 73, 2: 69, 3: 68, 4: 32, 5: 36, 6: ...
1       {8: 8, 9: 4, 10: 4, 11: 4, 12: 3, 13: 15, 14: ...
2       {25: 9, 26: 7, 27: 7, 28: 3, 29: 4, 30: 10, 31...
3       {44: 7, 1781: 1, 1179: 1, 112: 1, 508: 3, 581:...
4       {54: 14, 1496: 2, 2288: 1, 2315: 1, 3432: 1, 4...
                              ...                        
5874    {1709: 1, 409: 1, 170: 1, 662: 1, 253: 1, 255:...
5875    {2368: 1, 300: 1, 1385: 1, 2412: 1, 2059: 1, 1...
5876    {1827: 1, 869: 1, 2157: 1, 335: 1, 1931: 1, 21...
5877    {1038: 1, 701: 2, 495: 1, 353: 1, 189: 1, 484:...
5878                                             {694: 2}
Name: count_by_products, Length: 5879, dtype: object

W praktyce to oznacza, że  `customer_id=0` kupił np. `product_id=0` - 89 razy, `product_id=1` - 73 razy, `product_id=3` - 68 razy itd.


Fajnie byłoby nam wypakować słownik do osobnych kolumn, na szczęście da się to zrobić bardzo łatwo. ⌛️ Tylko poczekaj chwilkę na wykonanie.

In [13]:
df_count_products = df_customers["count_by_products"].apply(pd.Series).fillna(-1)
df_count_products.columns = ["product_{}".format(x) for x in df_count_products.columns]

df_count_products.head(5)

product_0  product_1  product_2  product_3  product_4  product_5  \
0       89.0       73.0       69.0       68.0       32.0       36.0   
1        6.0       -1.0        1.0       -1.0       -1.0       -1.0   
2       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   
3       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   
4       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   

   product_6  product_7  product_8  product_9  ...  product_4216  \
0       34.0       35.0       18.0       -1.0  ...          -1.0   
1       -1.0       -1.0        8.0        4.0  ...          -1.0   
2       -1.0       -1.0       -1.0       -1.0  ...          -1.0   
3       -1.0       -1.0       -1.0       -1.0  ...          -1.0   
4       -1.0       -1.0       -1.0       -1.0  ...          -1.0   

   product_4219  product_4220  product_4222  product_4226  product_4231  \
0          -1.0          -1.0          -1.0          -1.0          -1.0   
1          -1.0          -1.0          -1.0          -1.0          -1.0   
2          -1.0          -1.0          -1.0          -1.0          -1.0   
3          -1.0          -1.0          -1.0          -1.0          -1.0   
4          -1.0          -1.0          -1.0          -1.0          -1.0   

   product_4236  product_4237  product_4238  product_4239  
0          -1.0          -1.0          -1.0          -1.0  
1          -1.0          -1.0          -1.0          -1.0  
2          -1.0          -1.0          -1.0          -1.0  
3          -1.0          -1.0          -1.0          -1.0  
4          -1.0          -1.0          -1.0          -1.0  

[5 rows x 3878 columns]

Mamy teraz osobny `dataframe` z 3878 kolumnami. Każda kolumna to jest `product_id`. Wartość -1 oznacza brak informacji, czyli  klient nie kupował tego produktu, natomiast każda inna liczba (większa niż 0) oznacza, ile razy ten produkt był kupiony (to jest trochę optymistyczne zdanie, ale na razie możemy przyjąć, że jest ok).

☝️ Zwróć uwagę, że pojawiła się linijka zmiany nazwy produktu. `df_count_products.columns = ["product_{}".format(x) for x in df_count_products.columns]` W tym miejscu robimy bardzo prostą rzeczy, zamiast samego ID, np. 0 czy 2, dodajemy przedrostek, aby kolumna nazywała się  `product_0` czy `product_2`. To nam przyda się za chwilę.


### Następny krok

Połączmy `df_count_products` z `df_customers` i wytrenujmy model. Aby to połączyć użyjemy konstrukcji `concat()`. Wynikiem będzie to, że dokleimy to "z boku". Innymi słowy liczba wierszy zostanie taka sama, liczba kolumn powiększy się o 3878 😱. Model pewnie ucieszy się, będzie miał, co optymalizować.

Łączymy!

In [14]:
df_customers = pd.concat([df_customers, df_count_products], axis=1)
df_customers.shape

(5879, 3887)

Tak jak spodziewaliśmy się, liczba wierszy została taka sama - 5879 (bo liczba wierszy to są klienci, nowych klientów nie przybyło). Liczba kolumn wzrosła nam do 3887. Możesz o tym pomyśleć, że teraz każdy klient posiada 3887 cech, czyli charakterystyk.

## 🤖 Czas na trenowanie modelu

Najpierw sprawdźmy, czy jakość modelu się nam poprawiła.

Uwaga! Model potrzebuje czasu na trenowanie, więc możesz spokojnie przez chwilę odpocząć. Model wykonuje "prostą" i powtarzalną pracę, natomiast po Twojej stronie jest myślenie koncepcyjnie. Przyzwyczajaj się, to jest przyszłość, która już do nas przyszła :)

In [15]:
#df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)

train_and_get_scores(model, X, y)

(0.8474117817339701, 0.08924212099402075)

Jakość modelu poprawiła się, mamy już ~85%. 

Przypomnę, że same drzewa decyzyjne dały nam 77%. XGBoost na podstawowych cechach dał nam 81% i teraz mamy 85%.

Może się wydawać, że to słabo, bo mieliśmy 99,9%. Chcę taki wynik znowu! 

Być może to jest przykra wiadomość, ale często model "stopuje" się właśnie w okolicy 60-70% i potem trzeba naprawdę postarać się, aby wycisnąć więcej. No chyba, że użyjemy oczywistych zależności, ale to jest mało ciekawe. Lepiej jednak pozwolić modelowi znaleźć coś, co może zaskoczyć.

Wytrenujmy model teraz niezależnie, aby można było zajrzeć i sprawdzić, co dla niego jest ważne.

In [16]:
model.fit(X, y)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=50, n_jobs=0, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [17]:
eli5.show_weights(model, feature_names=feats, top=50)

No teraz już przynajmniej mamy na co popatrzeć :). Widać, że ważność produktu `product_545` oraz `product_99` jest ponad `count_orders`. Oczywiście, tak jak mówiłem, tutaj trzeba ostrożnie wyciągać wnioski. Natomiast to, co jest jednoznaczne to to, że `product_id=545` zasłużył na to, aby zwrócić na niego uwagę 🥳. To sprawdźmy :).

In [18]:
customer_ids_by_product = set(df[ df["product_id"] == 545 ]["customer_id"].unique())
len(customer_ids_by_product )

168

Ten produkt kupiło nie tak wielu klientów, tylko 168. Sprawdźmy, ilu z nich należy do segmentu `most_revenue_customer`.

In [19]:

df_customers[ df_customers.customer_id.isin(customer_ids_by_product) ]["most_revenue_customer"].mean()

0.4166666666666667

No widzimy, że mamy 42% klientów, którzy kupili co najmniej `product_id=545` i należą do segmentu `most_revenue_customer`.

Przypomnę, że normalnie mamy 22% klientów, którzy należą do segmentu `most_revenue_customer`.

In [20]:
df_customers["most_revenue_customer"].mean()

0.225208368770199

Możemy na szybko postawić hipotezę (do zbadania) -  jeśli klient kupił `product_id=545` to prawdopodobieństwo, że będzie należał do segmentu `most_revenue_customer` prawie podwaja się (od 23% do 42%). Brzmi jako "aha moment". Tę hipotezę warto jeszcze zbadać dokładniej, ale jeżeli ona faktycznie działa, to oznacza, że warto sprawić, aby więcej ludzi kupowało właśnie `product_id=545`.

#### Sprawdźmy, ile kosztuje ten produkt. Może to jakiś luksus ⚜️.

In [21]:
df[ df.product_id == 545 ]["price_unit"].value_counts()

125    126
39     103
42       5
246      3
106      1
Name: price_unit, dtype: int64

Z ceną tego produktu bywa różnie, ale najczęściej cena jednostkowa, to jedynie 1,25 lub 3,9 😱. Taki ciekawy wniosek udało się znaleźć w tak łatwy sposób.

## Zadanie 5.1


Twoim zadaniem jest zrobić podobną analizę dla produktu.


### 💡 Podpowiedzi: 

1. Przygotuj `df_products`
2. Dodaj więcej cech, przed dodaliśmy `product_id`, ale możesz zrobić symetryczne lub wymyśl coś innego (może `order_id`) lub jeszcze coś innego, jest sporo pomysłów :) 
3. Wytrenuj model. 
4. Zobacz, co dla modelu było najbardziej istotne.

In [42]:
df_products = (
    df[ ["price_total", "product_id"] ]
    .groupby("product_id")
    .agg("sum")
    .reset_index()
    .sort_values(by="price_total", ascending=False)
    .rename(columns={"price_total": "product_price_total"})
)


df_products["cumsum"] = df_products["product_price_total"].cumsum()
value_80prc = int(df["price_total"].sum() * 0.8)
df_products["most_revenue_product"] = df_products["cumsum"] < value_80prc


top_products = set(df_products[ df_products["most_revenue_product"] ]["product_id"].unique())

del df_products
gc.collect()

def feature_engineering(df):
    
    def counter(vals):
        cntr = Counter()
        cntr.update(vals)
        return cntr
    
    df_products = (
        df
        .groupby("product_id")
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_customers=("customer_id", lambda x: len(set(x))),
            count_by_customers=("customer_id", lambda x:  counter(x) ),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_products


def get_feats(df_products, black_list=["most_revenue_product"]):
    feats = list(df_products.select_dtypes([np.number, bool]).columns)
    return [x for x in feats if x not in black_list]

def get_X_y(df_products, top_products, feats):
    df_products["most_revenue_product"] = df_products["product_id"].map(lambda x: x in top_products)
    
    X = df_products[feats].values
    y = df_products["most_revenue_product"].values
    
    return X, y


def train_and_get_scores(model, X, y, scoring="accuracy", cv=5):

    scores = cross_val_score(model, X, y, scoring=scoring, cv=cv)
    return np.mean(scores), np.std(scores)


df_products  = feature_engineering(df)
df_products.head()

product_id  count_orders  count_unq_customers  \
0           0          5028                 1494   
1           1           614                  293   
2           2           736                  320   
3           3          1475                  706   
4           4           601                  324   

                                  count_by_customers  sum_quantity  \
0  {0: 89, 15: 36, 19: 19, 23: 5, 43: 2, 46: 3, 6...         97329   
1  {0: 73, 192: 4, 329: 4, 75: 6, 380: 3, 443: 2,...          4638   
2  {0: 69, 122: 2, 132: 10, 191: 2, 292: 8, 293: ...          7160   
3  {0: 68, 15: 3, 28: 2, 34: 2, 40: 3, 42: 11, 68...         16707   
4  {0: 32, 22: 4, 64: 2, 75: 3, 89: 2, 106: 5, 11...          3804   

   sum_price_unit  sum_price_total  count_unq_countries  prob_canceled  
0         1525206         26097687                   22       0.025179  
1          248166          1650981                    6       0.028358  
2          261086          2267898                   11       0.016753  
3          742149          5850395                   14       0.009699  
4          552166          2979279                   17       0.037994

In [43]:
df_products["count_by_customers"]


0       {0: 89, 15: 36, 19: 19, 23: 5, 43: 2, 46: 3, 6...
1       {0: 73, 192: 4, 329: 4, 75: 6, 380: 3, 443: 2,...
2       {0: 69, 122: 2, 132: 10, 191: 2, 292: 8, 293: ...
3       {0: 68, 15: 3, 28: 2, 34: 2, 40: 3, 42: 11, 68...
4       {0: 32, 22: 4, 64: 2, 75: 3, 89: 2, 106: 5, 11...
                              ...                        
3873                                            {4395: 1}
3874                      {327: 1, 580: 1, 150: 1, 47: 1}
3875                                       {47: 1, 68: 1}
3876                                             {448: 1}
3877                                             {398: 4}
Name: count_by_customers, Length: 3878, dtype: object

In [44]:
df_count_customers = df_products["count_by_customers"].apply(pd.Series).fillna(-1)
df_count_customers.columns = ["customer_{}".format(x) for x in df_count_customers.columns]

df_count_customers.head(5)

customer_0  customer_1  customer_2  customer_3  customer_4  customer_5  \
0        89.0         6.0        -1.0        -1.0        -1.0         4.0   
1        73.0        -1.0        -1.0        -1.0        -1.0        -1.0   
2        69.0         1.0        -1.0        -1.0        -1.0         3.0   
3        68.0        -1.0        -1.0        -1.0        -1.0         1.0   
4        32.0        -1.0        -1.0        -1.0        -1.0        -1.0   

   customer_6  customer_7  customer_8  customer_9  ...  customer_5869  \
0        -1.0        -1.0         5.0        -1.0  ...           -1.0   
1        -1.0        -1.0         1.0        -1.0  ...           -1.0   
2        -1.0        -1.0         1.0        -1.0  ...           -1.0   
3        -1.0        -1.0         3.0        -1.0  ...           -1.0   
4        -1.0        -1.0         9.0        -1.0  ...           -1.0   

   customer_5870  customer_5871  customer_5872  customer_5873  customer_5874  \
0           -1.0           -1.0           -1.0           -1.0           -1.0   
1           -1.0           -1.0           -1.0           -1.0           -1.0   
2           -1.0           -1.0           -1.0           -1.0           -1.0   
3           -1.0           -1.0           -1.0           -1.0           -1.0   
4           -1.0           -1.0           -1.0           -1.0           -1.0   

   customer_5875  customer_5876  customer_5877  customer_5878  
0           -1.0           -1.0            1.0           -1.0  
1           -1.0           -1.0           -1.0           -1.0  
2           -1.0           -1.0           -1.0           -1.0  
3           -1.0           -1.0           -1.0           -1.0  
4           -1.0            1.0           -1.0           -1.0  

[5 rows x 5879 columns]

In [53]:
df_products = pd.concat([df_products, df_count_customers], axis=1)
df_products.shape

(3878, 5889)

In [48]:
df_products = feature_engineering(df)
feats = get_feats(df_products, black_list=["most_revenue_product", "sum_price_total"])
X, y = get_X_y(df_products, top_products, feats)
model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)

train_and_get_scores(model, X, y)

(0.8633205853009643, 0.07143733771897691)

In [49]:
model.fit(X, y)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=50, n_jobs=0, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [50]:
eli5.show_weights(model, feature_names=feats)

### Wstępnie wnioski:
#### TOP3 
- `count_orders` brzmi jako najbardziej istotna cecha
- `sum_price_unit`
- `sum_quantity`

Wyglada na to, że cecha `count_unq_countries` jest mało przydatna

In [56]:
# df_products = feature_engineering(df)
feats = get_feats(df_products, black_list=["most_revenue_product", "sum_price_total"])
X, y = get_X_y(df_products, top_products, feats)
model = xgb.XGBClassifier(max_depth=5, n_estimators=50, learning_rate=0.3)

train_and_get_scores(model, X, y)

(0.8749321582973062, 0.08523619775762692)

In [58]:
model.fit(X, y)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=50, n_jobs=0, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [59]:
eli5.show_weights(model, feature_names=feats, top = 50)

Sprawdźmy `customer_1911`

In [62]:
product_ids_by_customer = set(df[ df["customer_id"] == 1911 ]["product_id"].unique())
len(product_ids_by_customer )

256

Ten klient zakupił 256 produktów. Sprawdźmy ile z tych produktów należy do segmentu `most_revenue_product`

In [64]:
df_products[ df_products.product_id.isin(product_ids_by_customer) ]["most_revenue_product"].mean()

0.6484375

Niemalże 65% produktów zakupionych przez customer_1911 znajduje się w segmencie `most_revenue_product`

In [65]:
df[ df.customer_id == 1911 ]["price_unit"].value_counts()

42      41
125     41
85      35
375     35
165     34
295     23
195     19
495     11
210     11
65       8
254      7
169      5
415      5
39       4
425      4
145      4
1695     4
325      3
83       3
19       3
595      3
850      2
208      2
894      2
55       2
465      2
249      2
14       1
36       1
12       1
21       1
8        1
6        1
795      1
289      1
675      1
1495     1
825      1
225      1
1250     1
4        1
Name: price_unit, dtype: int64

`Na podstawie powyższych danych można stwierdzić że jest to bardzo zróżnicowany klient. Są produkty, które zakupił 41 razy, ale mające dość niską cenę oraz są zakupione pojedyncze produkty, które mają znacznie wyższą cenę niż te, które zostały kupione największą ilość razy. `

Jak **wykonasz** zadanie, należy zrobić:

1. Zrzut ekranu na którym widać rozwiązania, wtedy dostaniesz bonus :)
2. Wrzucić rozwiązanie na slacku do pokoju **[#dwthon_day5_done](https://bit.ly/3teY3l8)**
3. Dostać bonus za dobrą robotę 💪 i jutro będzie kolejna porcja.

*Uwaga! Jeśli masz problem z jakimś zadaniem, czegoś nie wiesz, to pamiętaj, że możesz zadawać pytania! Na tym polega nauka :) Pytania związane z 2 dniem DWthon zadawaj w kanale [#dwthon_day5](https://bit.ly/3qEBEMv)*.

## 🧠 Włączmy myślenie krytycznie


Ta cała nasza analiza była oparta na `price_total`, czyli przychodach. Co ze zwrotami? Być może nastąpił czas nieco bardziej wejść do króliczej nory?

Na co jeszcze warto zwrócić krytyczną uwagę?  Podpowiem, że jeszcze jest sporo rzeczy ;)

## 😇 Włączmy myślenie kreatywne

Pomyśl, jak możesz wykorzystać tę wiedzę, którą tutaj udało Ci się zdobyć?

Najlepiej, aby padły przykłady z Twojego podwórka. Jeśli nie masz (ale pomyśl, bo to tylko może się wydawać, że nie ma) to pomyśl o swoim otoczeniu.

Od razu powiem, że nie chodzi tylko o e-commerce. To może być dowolna branża. Reguła 80/20 jest uniwersalna. To samo dotyczy "aha-momentów".

## 🤝🗣️ Pomyśl o kooperacji i komunikacji

#### ☝️ Podziel się swoimi przemyśleniami na slacku **[#dwthon_day5_ideas](https://bit.ly/3qFcb5R)**. Jeśli masz pomysł, ale nie masz kodu - poproś na slacku o pomoc, aby ktoś pomógł Ci napisać kod. Natomiast jeśli masz kod, to koniecznie podziel się. Dzięki temu ktoś inny będzie w stanie łatwo powtórzyć i podzielić się swoimi przemyśleniami. Dzięki temu każdy zyskuje :).


In [ ]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/qC2MZcjGfcM" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## 👇👇👇 Twój Feedback i delta rozwoju 👇👇👇

Daj mi znać koniecznie, co sprawiło Ci przyjemność, co trudność oraz czego udało się nauczyć w tej krótkiej [ankiecie](https://bit.ly/3rEnOLD). Pozwoli mi ona ulepszyć zadania, a Tobie lepiej trakować swój rozwój: win-win 😇

In [ ]:
%%html

<div data-tf-widget="zUbU39eg" data-tf-iframe-props="title=Feedback after 5 day" data-tf-medium="snippet" style="width:100%;height:400px;"></div><script src="//embed.typeform.com/next/embed.js"></script>


## [DW Club](https://bit.ly/3rGtHtv) ❤️

I oczywiście to jest bardzo dobry moment, aby właśnie teraz dołączyć się >> **[DW Club](https://bit.ly/3rGtHtv)** << bo **zyskujesz wiele** za małą cenę 😎. Zobacz:
1. dostęp do tego DWthon przez cały czas, dla Ciebie to oznacza, że możesz wrócić do tematu po jego zakończeniu wtedy kiedy tylko chcesz
2. gotowe środowisko, które czeka 24/7 na Ciebie, działaj wtedy jak masz czas, nie musisz nic instalować oraz utrzymywać przez cały czas (biblioteki cały czas zmieniają się)
3. gotowe rozwiązania (oprócz DWthon, już tam sporo rzeczy zebrało się - np. master class, dwgym itd), możesz inspirować się, i też traktować to jako "ściągawka", którą masz pod ręką jak tylko go potrzebujesz
4. determinację do działania (razem zwykle jest łatwiej), trudniej będzie przekładać na potem ;)
5. oszczędzaj czasem razem z nami, tym bardziej jak go masz bardzo mało, sporo informacji filtrujemy przez nasze doświadczenia
6. praktyczna wiedza z praktykującym, uczę Cię osoba, która cały czas realizuje i wdraża prawdziwy projekty. Zrobisz duży postęp, po prostu obserwując sposób myślenia i praktykując
7. przyjazna społeczność, czyli spokojnie możesz pytać, rozważać, dyskutować (bardzo dbamy o to, aby szacunek był na wysokim poziomie)


Co najmniej jedne nowe zadanie miesięcznie... I jeszcze wiele innych korzyści dla Ciebie

### >> [Dołączasz się teraz](https://bit.ly/3rGtHtv) <<?

## 💪 Pochwal się światu, że DWthon za Tobą 💪 

Udało Ci się! Jesteś na mecie, koniecznie powiedz o tym innym. Być może kogoś to zmotywuje do działania, a dodatkowo warto mówić głośno o nowych umiejętnościach i doświadczeniach, bo nigdy nie wiesz, jaka furtka może się otworzyć ;) 

Pamiętaj o dodaniu tagów: `#dwclub` `#dwthon` `#hackoutsidethebox` `#machinelearning` `#datascience` ;) 


Śmiało nas oznaczaj, będzie nam miło :))

# 😇 Śledź nas w social media

- [youtube](https://www.youtube.com/channel/UCRLds2lEuh7cfkvTR4I_ddQ)
- [facebook](https://www.facebook.com/groups/dataworkshop/)
- [linkedin](https://www.linkedin.com/company/dataworkshop)
- [instagram](https://www.instagram.com/data.vladimir/)
- [telegram](https://t.me/dataworkshop)


Też śmiało wysyłaj zaproszenie na [LinkedIn Vladimira](https://www.linkedin.com/in/vladimiralekseichenko/). Tylko się ucieszę z tego powodu ;)

## 🤖 Polecam obejrzyć webinarium: [10 najpopularniejszych błędów w uczeniu maszynowym](https://bit.ly/3qKQbqg) 🤖 